### PR TITLE
DPTP-2199: Allow setting multi-stage params externally

### PR DESCRIPTION
**Pull request by emilvberglind**
With [DPTP-2199](https://issues.redhat.com/browse/DPTP-2199) we want to allow setting multi-stage parameters externally through the ci-operator

### CODE REVIEW

File: cmd/ci-operator/main_test.go [hunk 1]
```diff
@@ -543,3 +543,119 @@ func TestExcludeContextCancelledErrors(t *testing.T) {
...
+			inputParams:    stringSlice{[]string{"PARAM1", "PARAM2"}},
+			expectedParams: map[string]string{},
+			expectedErrs: []string{
+				"could not parse multi-stage-param: PARAM1 is not in the format key=value",
+				"could not parse multi-stage-param: PARAM2 is not in the format key=value",
+			},
+		},
+	}
+
+	for _, tc := range testCases {
```
Comment (petr-muller) [inline cmd/ci-operator/main_test.go:619]: Check out how other tests use subtests with `t.Run` and `t.Parallel` (https://golang.org/pkg/testing/) to improve the experience for whoever will run this.

File: cmd/ci-operator/main_test.go [hunk 2]
```diff
@@ -543,3 +543,119 @@ func TestExcludeContextCancelledErrors(t *testing.T) {
...
+				}
+			}
+
+			if test.MultiStageTestConfiguration != nil {
+				for name, val := range test.MultiStageTestConfiguration.Environment {
+					actualParams[name] = val
+				}
+			}
+		}
+		if diff := cmp.Diff(tc.expectedParams, actualParams); diff != "" {
```
Comment (petr-muller) [inline cmd/ci-operator/main_test.go:645]: I think you can cut the implementation some slack and only test for this when there's no error - it is a convention that when `err` is not nil then the other return values can contain garbage, and some implementations take advantage of it, so the outputs only make sense to test when there was no error.

File: cmd/ci-operator/main.go [hunk 3]
```diff
@@ -404,6 +405,8 @@ type options struct {
 
 	hiveKubeconfigPath string
 	hiveKubeconfig     *rest.Config
+
+	multiStageParamsSlice stringSlice
```
[the remark this answers is not in the document]
Comment (emilvberglind) [reply cmd/ci-operator/main.go:409]: Yeah, that's better - I suck at naming things

File: cmd/ci-operator/main.go [hunk 4]
```diff
@@ -651,6 +656,56 @@ func (o *options) Complete() error {
...
+		// for any multi-stage tests, go ahead and inject the passed-in parameters. Note that parameters explicitly passed
+		// in to ci-operator will take precedence.
+		for _, test := range o.configSpec.Tests {
+			var testEnvironment api.TestEnvironment
+			if test.MultiStageTestConfigurationLiteral != nil {
+				if test.MultiStageTestConfigurationLiteral.Environment == nil {
+					test.MultiStageTestConfigurationLiteral.Environment = make(api.TestEnvironment)
+				}
+				testEnvironment = test.MultiStageTestConfigurationLiteral.Environment
+			} else if test.MultiStageTestConfiguration != nil {
```
[the remark this answers is not in the document]
Comment (emilvberglind) [reply cmd/ci-operator/main.go:694]: Oh ok, cool, I wasn't sure about that so erred on the side of caution. I'll drop that second clause

File: cmd/ci-operator/main.go [hunk 5]
```diff
@@ -465,6 +468,8 @@ func bindOptions(flag *flag.FlagSet) *options {
 
 	flag.StringVar(&opt.hiveKubeconfigPath, "hive-kubeconfig", "", "Path to the kubeconfig file to use for requests to Hive.")
 
+	flag.Var(&opt.multiStageParamsSlice, "multi-stage-param", "One or more environment parameters that will be passed down to the multi-stage steps")
```
[the remark this answers is not in the document]
Comment (emilvberglind) [reply cmd/ci-operator/main.go:471]: How about something like: `A repeatable option where one or more environment parameters that will be passed down to the multi-stage steps, in the format NAME=VAL. e.g --multi-stage-param PARAM1=VAL1 --multi-stage-param PARAM2=VAL2.`

**Review by petr-muller — APPROVED** [review]

**Pull request merged**